R/generate.R

```diff
@@ -0,0 +1,72 @@
+int_types <- c("integer", "smallint")
+logi_types <- c("boolean")
+num_types <- c("numeric", "real")
+date_types <- c("date", "timestamp without time zone", "timestamp with time zone")
+
+validate_type <- quote(stopifnot(typeof(x) == y))
+validate_inherits <- quote(stopifnot(inherits(x, y)))
+validate_nulls <- quote(stopifnot(!any(is.na(x))))
+
+build <- function(table, env = topenv()) {
+  args <- do.call(build_args, table)
+  names(args) <- table$column_name
+  cols <- lapply(table$column_name, as.name)
+  names(cols) <- table$column_name
+  expr <- as.call(c(list(quote(data.frame)),
+                    cols,
+                    list(stringsAsFactors = FALSE)))
+
+  body <- c(
+    unlist(Map(function(col_name, is_nullable, r_type) {
+      env <- list(x = as.name(col_name), y = r_type)
+      c(if (r_type == "POSIXct") substitute_(validate_inherits, env),
+        if (r_type != "POSIXct") substitute_(validate_type, env),
+        if (is_nullable == "NO") substitute_(validate_nulls, env))
+    },
+               table$column_name, table$is_nullable, table$r_type)),
+    list(expr))
+
+  as_function(args, body, env)
+}
+
+build_args <- function(is_nullable, r_type, ...) {
+  res <- rep(alist(. =), length(is_nullable))
+  res[is_nullable == "YES" & r_type == "integer"] <- alist(. = NA_integer_)
```

Comment:
@richfitz is there a way to declare a single argument, as opposed to an `alist`?

Reply:
this weird construction exists to allow the names to be changed - normally you'd use it like `alist(a =, b = NULL)` or similar.  You can use `as.pairlist` but getting the empty arg there is hard otherwise (the function argument with no default).

I don't really follow the question though, just hoping this information bridges the gap